gpytorch/models/deep_gps/deep_gp.py

```diff
@@ -0,0 +1,173 @@
+from gpytorch.mlls import AddedLossTerm
+from gpytorch.likelihoods import GaussianLikelihood
+from gpytorch import settings
+from torch.nn.functional import softplus
+
+
+class NegativeKLDivergence(AddedLossTerm):
+    def __init__(self, variational_strategy):
+        self.variational_strategy = variational_strategy
+
+    def loss(self):
+        return -1 * self.variational_strategy.kl_divergence().sum()
+
+
+class AbstractDeepGPLayer(AbstractVariationalGP):
+    def __init__(self, variational_strategy, input_dims, output_dims, num_samples, last_layer=False):
+        """
+        Represents a layer in a deep GP where inference is performed via the doubly stochastic method of
+        Salimbeni et al., 2017. Upon calling, instead of returning a variational distribution q(f), returns samples
+        from the variational distribution.
+
+        See the documentation for __call__ below for more details below. Note that the behavior of __call__
+        will change to be much more elegant with multiple batch dimensions; however, the interface doesn't really
+        change.
+
+        Args:
+            - variational_strategy (VariationalStrategy): Strategy for changing q(u) -> q(f) (see other VI docs)
+            - input_dims (int): Dimensionality of input data expected by each GP
+            - output_dims (int): Number of GPs in this layer, equivalent to output dimensionality.
+            - last_layer (bool): True if this is to be the last layer in the deep GP, false otherwise.
+            - num_samples (int): Number of samples to draw from q(f) for returning
+        """
+        super(AbstractDeepGPLayer, self).__init__(variational_strategy)
+        self.input_dims = input_dims
+        self.output_dims = output_dims
+        self.num_samples = num_samples
+        self.last_layer = last_layer
+        self.register_added_loss_term("hidden_kl_divergence")
+
+    def forward(self, x):
+        raise NotImplementedError
+
+    def _reshape_input(self, inputs):
+        inputs = inputs.contiguous()
+        # Forward samples through the VariationalStrategy and return *samples* from q(f)
+        if inputs.dim() == 2:
+            # Assume new input entirely
+            inputs = inputs.unsqueeze(0)
+            inputs = inputs.expand(self.output_dims, inputs.size(-2), -1)
+        elif inputs.dim() == 3:
+            # Assume batch dim is samples, not output_dim
+            inputs = inputs.unsqueeze(0)
+            inputs = inputs.expand(self.output_dims, inputs.size(1), inputs.size(-2), -1)
+
+        return inputs
+
+    def __call__(self, inputs):
+        """
+        Forward data through this hidden GP layer.
+
+        If the input is 2 dimensional, we pass the input through each hidden GP, resulting in a `h x n` batch
+        Gaussian distribution. We then draw `s` samples from these Gaussians and reshape the result to be
+        `s x n x h` (e.g., h becomes the dimensionality of the output).
```

Comment:
I wonder - would it make sense for all layers to return a MultivariateNormal distribution object? __call__ would be designed to handle tensor inputs (default behavior) or distribution inputs (draw a sample from them).